source/common/quic/envoy_quic_client_stream.cc

```diff
@@ -228,6 +228,22 @@ void EnvoyQuicClientStream::OnStreamFrame(const quic::QuicStreamFrame& frame) {
   quic::QuicSpdyClientStream::OnStreamFrame(frame);
 }
 
+bool EnvoyQuicClientStream::OnStopSending(quic::QuicResetStreamError error) {
+  // Only called in IETF Quic to close read side.
+  ENVOY_STREAM_LOG(debug, "received STOP_SENDING with reset code={}", *this, error.internal_code());
+  bool end_stream_encoded = local_end_stream_;
+  // This call will close write.
+  if (!quic::QuicSpdyClientStream::OnStopSending(error)) {
+    return false;
+  }
+  if (!end_stream_encoded) {
```

Comment:
Good point. I think it was working before because the server would only STOP_SENDING after closing, but that cannot be assumed in general.

I'm not sure about OnStreamReset(). What would be broken if we leave it as-is?

Reply:
> I'm not sure about OnStreamReset(). What would be broken if we leave it as-is?

You need to check write_side_closed() before calling runResetCallbacks(), because OnStreamReset() only close read side.

Reply:
Updated to mirror server side. Also updated a test that was incorrect.

Reply:
Thanks for the additional tests!

Reply:
Does the IntegrationStreamDecoder lifetime issue still exist now?

Reply:
Let me try to push a commit and see? I can't reproduce the sanitizer issue locally for some reason, only on CI.

Reply:
Yes, the lifetime issue still exists, see this log https://dev.azure.com/cncf/envoy/_build/results?buildId=115672&view=logs&jobId=4e2e4662-2d21-5726-b4bc-9ff5cd6a179b&j=4e2e4662-2d21-5726-b4bc-9ff5cd6a179b&t=a4553756-733d-5e3b-e02d-5aef5f5e9bee. I'll revert the test commit.

Reply:
Sorry, the link doesn't load the test result for me. What's the msan error?

Reply:
Argh, the log disappeared for me, too :( It's msan violation, access to deleted IntegrationStreamDecoder from Reset callback.

Reply:
The test is `FaultyFilterWithConnect`.

Reply:
It's a msan failure in Linux, can you repro the test locally? It'll great if you can explain a bit more about what the life cycle difference between HTTP/3 and HTTP/2 that caused this failure. It's unclear to me if this is a test setup issue or a newly introduced bug in quic client stream objects life cycle.

Reply:
Let me try to undo test changes and instead just close the request. I think that is more reasonable.

Reply:
Sample error https://gist.github.com/kyessenov/016643c55bf2c34863954c76b9f805dd

Reply:
The root cause was determined to be the early test termination before client had a chance to read STOP_SENDING and fully reset before the connection closure. Thank you @danzh2010 for assistance.

Reply:
You're welcome! Side note: can https://github.com/envoyproxy/envoy/pull/22802/commits/b18d968bffbc56024ac775f109b9d3d750daa976 be applied forward now?

Reply:
I actually removed all the changes to the test framework, so that commit is no longer necessary. The PR is much smaller now.